Sources/GDBRemote/Structures.cpp

```diff
@@ -79,60 +79,75 @@ bool Feature::parse(std::string const &string) {
 //
 
 bool ProcessThreadId::parse(std::string const &string, CompatibilityMode mode) {
-#define CHECK_AND_RESET(RESULT, RESET_VAL)                                     \
-  do {                                                                         \
-    if (errno != 0) {                                                          \
-      (RESULT) = (RESET_VAL);                                                  \
-      return false;                                                            \
-    }                                                                          \
-  } while (0)
-
-  pid = kAllProcessId;
-  tid = kAllThreadId;
+  this->pid = kAllProcessId;
+  this->tid = kAllThreadId;
 
   if (string.empty())
     return false;
 
+  ProcessId new_pid = kAllProcessId;
+  ThreadId new_tid = kAllThreadId;
+
   errno = 0;
 
-  if (mode == kCompatibilityModeGDB ||
-      mode == kCompatibilityModeGDBMultiprocess) {
-    if (string[0] == 'p') {
-      char *eptr;
-      pid = strtoul(string.c_str() + 1, &eptr, 16);
-      CHECK_AND_RESET(pid, kAllProcessId);
+  const char *str = string.c_str();
+  char *eptr = nullptr;
+  switch (mode) {
+  case kCompatibilityModeGDB:
+  case kCompatibilityModeGDBMultiprocess:
+    if (*str == 'p') {
+      str += 1;
+      new_pid = std::strtoul(str, &eptr, 16);
+      if (str == eptr || errno != 0)
+        return false;
+
       if (*eptr++ == '.') {
-        tid = strtoul(eptr, nullptr, 16);
-        CHECK_AND_RESET(tid, kAllThreadId);
+        str = eptr;
+        new_tid = std::strtoul(str, &eptr, 16);
+        if (str == eptr || errno != 0)
+          return false;
       }
     } else {
-      pid = std::strtoul(string.c_str(), nullptr, 16);
-      CHECK_AND_RESET(pid, kAllProcessId);
+      new_pid = std::strtoul(str, &eptr, 16);
+      if (str == eptr || errno != 0)
+        return false;
```

Comment:
This might be a nice lambda:

```c++
auto parse_hex = [&eptr](const char *str, unsigned long &value) -> bool {
  uint64_t value = std::strtoul(str, &eptr, 16);
  return !(str == eptr || errno);
};
```

That would allow us to just do `if (!parse_hex(str, pid)) return false;` rather than repeating the pattern.

Reply:
Ok yeah. I'm slowly learning newer C++.